mlir-tensorrt/compiler/lib/Compiler/OptionsProviders.cpp

```diff
@@ -0,0 +1,64 @@
+//===- OptionsProviders.cpp -------------------------------------*- C++ -*-===//
+//
+// SPDX-FileCopyrightText: Copyright 2024 NVIDIA CORPORATION & AFFILIATES.
+// All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//===----------------------------------------------------------------------===//
+///
+/// Data structures and functions for manipulating compiler options.
+///
+//===----------------------------------------------------------------------===//
+#include "mlir-tensorrt/Compiler/OptionsProviders.h"
+#include "cuda_runtime_api.h"
+#include "llvm/Support/Error.h"
+
+// TODO (pranavm): Check if we can just reuse `DeviceInfo.cpp`?
+llvm::Error mlirtrt::compiler::DeviceOptions::inferDeviceOptionsFromHost() {
+  cudaDeviceProp properties;
+  cudaError_t err = cudaGetDeviceProperties(&properties, 0);
+  if (err != cudaSuccess)
+    return llvm::createStringError("failed to get cuda device properties");
+
+  int ccMajor = 0;
+  int ccMinor = 0;
+  err = cudaDeviceGetAttribute(
+      &ccMajor, cudaDeviceAttr::cudaDevAttrComputeCapabilityMajor, 0);
+  if (err != cudaSuccess)
+    return llvm::createStringError(
+        "failed to get cuda device compute capability");
+  err = cudaDeviceGetAttribute(
+      &ccMinor, cudaDeviceAttr::cudaDevAttrComputeCapabilityMinor, 0);
+  if (err != cudaSuccess)
+    return llvm::createStringError(
+        "failed to get cuda device compute capability");
+
+  // We want SM version as a single number.
+  int64_t smVersion = ccMajor * 10 + ccMinor;
+  this->computeCapability = smVersion;
+  this->maxSharedMemoryPerBlockKb = properties.sharedMemPerBlock / 1024;
+  this->maxRegistersPerBlock = properties.regsPerBlock;
+  return llvm::Error::success();
+}
+
+llvm::Error mlirtrt::compiler::DeviceOptions::finalize() {
+  if (shouldInferFromHost) {
+    // TODO (pranavm): How to check whether options were provided?
```

Comment:
Not sure how to validate the options here. I don't know if we have access to the actual CLI options